File: backend/app/engine/stockfish_engine.py
"""Thin wrapper around a real Stockfish process (via python-chess's UCI client).

This module is the ONLY place that talks to Stockfish. It knows nothing about
game history, PGN, or persistence — it only ever answers "given this
position, at this difficulty, what move should the engine play (or how good
is this position)". Game state and rules live in app.game, never here.
"""
from __future__ import annotations

import random
import time
from dataclasses import dataclass

import chess
import chess.engine

from app.core.config import get_settings
from app.engine.difficulty import DifficultyProfile, get_profile


@dataclass
class EngineMoveResult:
    move: chess.Move
    thinking_time_ms: int
    was_blunder: bool  # True if a deliberately sub-optimal move was chosen


class EngineUnavailableError(RuntimeError):
    """Raised when the Stockfish binary cannot be launched."""


class StockfishEngine:
    """Owns one Stockfish subprocess for the lifetime of the instance.

    Not thread-safe for concurrent `best_move` calls on the same instance;
    callers (e.g. the game service) should serialize access per game.
    """

    def __init__(self, stockfish_path: str | None = None) -> None:
        settings = get_settings()
        self._path = stockfish_path or settings.stockfish_path
        self._max_movetime_ms = settings.engine_max_movetime_ms
        self._max_depth = settings.engine_max_depth
        try:
            self._engine = chess.engine.SimpleEngine.popen_uci(self._path)
        except (FileNotFoundError, OSError) as exc:
            raise EngineUnavailableError(
                f"Could not start Stockfish at '{self._path}'. "
                "Install it (e.g. `apt-get install stockfish`) or set COACH_STOCKFISH_PATH."
            ) from exc

    def close(self) -> None:
        try:
            self._engine.quit()
        except Exception:
            pass

    def __enter__(self) -> "StockfishEngine":
        return self

    def __exit__(self, *_exc) -> None:
        self.close()

    def best_move(self, board: chess.Board, target_elo: int) -> EngineMoveResult:
        """Return the engine's chosen move for `board`, calibrated to `target_elo`."""
        profile = get_profile(target_elo)
        self._configure(profile)

        limit_kwargs: dict = {}
        depth = min(profile.depth, self._max_depth) if profile.depth else None
        movetime_s = min(profile.movetime_ms, self._max_movetime_ms) / 1000.0
        if depth is not None:
            limit_kwargs["depth"] = depth
        limit_kwargs["time"] = movetime_s
        limit = chess.engine.Limit(**limit_kwargs)

        started = time.monotonic()

        use_blunder = (
            profile.blunder_probability > 0
            and random.random() < profile.blunder_probability
        )

        if use_blunder and profile.blunder_multipv > 1:
            legal_count = board.legal_moves.count()
            multipv = min(profile.blunder_multipv, max(legal_count, 1))
            info_list = self._engine.analyse(board, limit, multipv=multipv)
            candidates = [info["pv"][0] for info in info_list if info.get("pv")]
            move = random.choice(candidates) if candidates else self._engine.play(board, limit).move
            was_blunder = len(candidates) > 1
        else:
            result = self._engine.play(board, limit)
            move = result.move
            was_blunder = False

        elapsed_ms = int((time.monotonic() - started) * 1000)
        return EngineMoveResult(move=move, thinking_time_ms=elapsed_ms, was_blunder=was_blunder)

    def evaluate(self, board: chess.Board, depth: int = 12) -> chess.engine.PovScore:
        """Static evaluation of a position at full strength (used later for analysis)."""
        self._configure(get_profile(2100))
        info = self._engine.analyse(board, chess.engine.Limit(depth=depth))
        return info["score"]

    def _configure(self, profile: DifficultyProfile) -> None:
        options: dict[str, object] = {"Skill Level": profile.skill_level}
        if profile.use_uci_limit_strength and profile.uci_elo is not None:
            options["UCI_LimitStrength"] = True
            options["UCI_Elo"] = profile.uci_elo
        else:
            options["UCI_LimitStrength"] = False
        self._engine.configure(options)

File: frontend/src/engine/stockfishEngine.ts
/**
 * Talks to a real Stockfish compiled to WebAssembly, running in a Web Worker.
 *
 * This is the ONLY module that speaks UCI. It knows nothing about game
 * history, PGN or storage — it answers two questions about a position:
 * "which move do you play at this difficulty?" (game play, Phase 1) and
 * "how do you assess it?" (analysis, Phase 2). Everything runs on the
 * device: no network, no server.
 */
import { getProfile } from "./difficulty";
import type { Score } from "../analysis/evaluation";

export interface EngineMoveResult {
  uci: string;
  thinkingTimeMs: number;
  wasBlunder: boolean; // a deliberately sub-optimal move was chosen
}

/** One candidate line returned by the engine, ranked 1 = best. */
export interface EngineLine {
  /** MultiPV rank, 1-based. */
  rank: number;
  /** Score from the SIDE TO MOVE's point of view, as UCI reports it. */
  score: Score;
  /** Principal variation, in UCI moves. */
  pv: string[];
  /** Depth actually reached for this line. */
  depth: number;
}

export interface AnalysisResult {
  lines: EngineLine[];
  depth: number;
  timeMs: number;
}

export interface AnalyseOptions {
  depth: number;
  movetimeMs: number;
  multiPv: number;
}

/** Where the engine assets live (copied out of node_modules by scripts/copy-engine.mjs). */
const ENGINE_URL = `${import.meta.env.BASE_URL}engine/stockfish.wasm.js`;

export class EngineUnavailableError extends Error {}

export class StockfishEngine {
  private worker: Worker;
  private listeners = new Set<(line: string) => void>();
  private readyPromise: Promise<void>;

  constructor() {
    try {
      this.worker = new Worker(ENGINE_URL);
    } catch (cause) {
      throw new EngineUnavailableError(`Could not start the chess engine from ${ENGINE_URL}`);
    }
    this.worker.onmessage = (event: MessageEvent) => {
      const line = typeof event.data === "string" ? event.data : "";
      if (line) for (const listener of [...this.listeners]) listener(line);
    };
    this.readyPromise = this.handshake();
  }

  /** Resolves once the engine has answered `uciok` + `readyok`. */
  ready(): Promise<void> {
    return this.readyPromise;
  }

  terminate(): void {
    this.listeners.clear();
    this.worker.terminate();
  }

  private send(command: string): void {
    this.worker.postMessage(command);
  }

  /** Waits until `predicate` accepts a line, feeding every line to `onLine`. */
  private waitFor(
    predicate: (line: string) => boolean,
    onLine?: (line: string) => void,
    timeoutMs = 60000,
  ): Promise<string> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.listeners.delete(listener);
        reject(new EngineUnavailableError("The chess engine stopped responding."));
      }, timeoutMs);

      const listener = (line: string) => {
        onLine?.(line);
        if (predicate(line)) {
          clearTimeout(timer);
          this.listeners.delete(listener);
          resolve(line);
        }
      };
      this.listeners.add(listener);
    });
  }

  private async handshake(): Promise<void> {
    this.send("uci");
    await this.waitFor((l) => l === "uciok" || l.startsWith("uciok"));
    this.send("isready");
    await this.waitFor((l) => l.startsWith("readyok"));
  }

  /**
   * Best move for `fen`, calibrated to `targetElo`.
   * `moves` is the UCI move list from the starting position, so the engine
   * sees repetitions (needed for threefold-repetition awareness).
   */
  async bestMove(fen: string, targetElo: number, randomFn: () => number = Math.random): Promise<EngineMoveResult> {
    await this.readyPromise;
    const profile = getProfile(targetElo);

    const wantsBlunder = profile.blunderProbability > 0 && randomFn() < profile.blunderProbability;
    const multiPv = wantsBlunder ? profile.blunderMultiPv : 1;

    this.send(`setoption name Skill Level value ${profile.skillLevel}`);
    this.send(`setoption name MultiPV value ${multiPv}`);
    this.send(`position fen ${fen}`);

    // Keep the deepest first-move seen for each MultiPV slot. Deeper
    // iterations overwrite shallower ones, so at `bestmove` this holds the
    // engine's final ranking.
    const candidatesByPv = new Map<number, string>();
    const startedAt = performance.now();

    this.send(`go depth ${profile.depth} movetime ${profile.movetimeMs}`);

    const bestmoveLine = await this.waitFor(
      (line) => line.startsWith("bestmove"),
      (line) => {
        if (!line.startsWith("info ") || !line.includes(" pv ")) return;
        const pvIndexMatch = line.match(/ multipv (\d+)/);
        const pvMatch = line.match(/ pv ([a-h][1-8][a-h][1-8][qrbn]?)/);
        if (!pvMatch) return;
        candidatesByPv.set(pvIndexMatch ? Number(pvIndexMatch[1]) : 1, pvMatch[1]);
      },
    );

    const thinkingTimeMs = Math.round(performance.now() - startedAt);
    const bestmove = bestmoveLine.split(/\s+/)[1];

    if (!bestmove || bestmove === "(none)") {
      throw new EngineUnavailableError("The chess engine returned no move.");
    }

    if (wantsBlunder) {
      const candidates = [...candidatesByPv.values()];
      if (candidates.length > 1) {
        const picked = candidates[Math.floor(randomFn() * candidates.length)];
        return { uci: picked, thinkingTimeMs, wasBlunder: picked !== bestmove };
      }
    }

    return { uci: bestmove, thinkingTimeMs, wasBlunder: false };
  }

  /**
   * Assesses a position at full strength (Phase 2 analysis).
   *
   * Unlike `bestMove`, this never weakens the engine: Skill Level is forced
   * back to 20, because an analysis must reflect the best judgement available,
   * not the difficulty the player chose to face.
   *
   * Scores come back from the SIDE TO MOVE's point of view, exactly as UCI
   * reports them; converting is the caller's explicit job (see
   * analysis/evaluation.ts).
   */
  async analyse(fen: string, options: AnalyseOptions): Promise<AnalysisResult> {
    await this.readyPromise;

    this.send("setoption name Skill Level value 20");
    this.send(`setoption name MultiPV value ${options.multiPv}`);
    this.send(`position fen ${fen}`);

    // Deeper iterations overwrite shallower ones for the same MultiPV slot,
    // so at `bestmove` this map holds the engine's final ranking.
    const linesByRank = new Map<number, EngineLine>();
    const startedAt = performance.now();

    this.send(`go depth ${options.depth} movetime ${options.movetimeMs}`);

    await this.waitFor(
      (line) => line.startsWith("bestmove"),
      (line) => {
        const parsed = parseInfoLine(line);
        if (parsed) linesByRank.set(parsed.rank, parsed);
      },
    );

    const lines = [...linesByRank.values()].sort((a, b) => a.rank - b.rank);
    return {
      lines,
      depth: lines.length ? Math.max(...lines.map((l) => l.depth)) : 0,
      timeMs: Math.round(performance.now() - startedAt),
    };
  }
}

/**
 * Parses one UCI `info` line into a candidate line.
 *
 * Returns null for the many info lines that carry no principal variation
 * (currmove, nps-only updates, ...). Exported for testing: getting this
 * parsing wrong silently corrupts every evaluation downstream.
 */
export function parseInfoLine(line: string): EngineLine | null {
  if (!line.startsWith("info ") || !line.includes(" pv ")) return null;

  const depthMatch = line.match(/ depth (\d+)/);
  const rankMatch = line.match(/ multipv (\d+)/);
  const cpMatch = line.match(/ score cp (-?\d+)/);
  const mateMatch = line.match(/ score mate (-?\d+)/);
  const pvMatch = line.match(/ pv (.+)$/);

  if (!pvMatch || (!cpMatch && !mateMatch)) return null;

  const pv = pvMatch[1].trim().split(/\s+/).filter((m) => /^[a-h][1-8][a-h][1-8][qrbn]?$/.test(m));
  if (pv.length === 0) return null;

  const score: Score = mateMatch
    ? { type: "mate", value: Number(mateMatch[1]) }
    : { type: "cp", value: Number(cpMatch![1]) };

  return {
    rank: rankMatch ? Number(rankMatch[1]) : 1,
    score,
    pv,
    depth: depthMatch ? Number(depthMatch[1]) : 0,
  };
}
